azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -273,6 +273,109 @@ def _handle_pipeline_event(self, event):
                 logger.warning("incoming pipeline event with no handler.  dropping.")
 
 
+# NOTE: This stage could be a candidate for being refactored into some kind of other
+# pipeline-related structure. What's odd about it as a stage is that it doesn't really respond
+# to operations or events so much as it spawns them on a timer.
+# Perhaps some kind of... Pipeline Daemon?
+class SasTokenRenewalStage(PipelineStage):
+    # Amount of time, in seconds, prior to token expiration, when the renewal process will begin
+    DEFAULT_TOKEN_RENEWAL_MARGIN = 120
```

Comment:
>DEFAULT_TOKEN_RENEWAL_MARGIN = 120 [](start = 4, length = 34)

why is this set here instead of setting this in the pipeline config with the renewal interval?  part of the purpose of this is to account for clock skew between the client and the server and it's not had to imagine needing a bigger margin if you know that your device clock is going to be inaccurate for some reason or another.

Reply:
to be clear are you saying this would be user input kwarg ?

---
In reply to: [427021277](https://github.com/Azure/azure-iot-sdk-python/pull/555#discussion_r427021277) [](ancestors = 427021277)

Reply:
yes, margin should be a config kwarg that the user can supply

---
In reply to: [428953125](https://github.com/Azure/azure-iot-sdk-python/pull/555#discussion_r428953125) [](ancestors = 428953125,427021277)